### ISSUE TITLE
Generic type parameters not propagated to names bound in match class patterns

### ISSUE THREAD

**Issue by Jayllyz** [astral-sh/ty#3879]
## Summary

When matching on a generic subclass using positional class patterns (via `__match_args__`), bound names are inferred as `object` instead of the concrete type parameter.

This worked silently in `0.0.51` because bound names were `Unknown`. Since `0.0.53` introduced \"Infer types for names bound in match patterns\" (#25940), the names are now correctly inferred — but for generic classes, the type parameter is not propagated and falls back to `object`, causing false-positive errors.

**Minimal reproduction:**

```python
from __future__ import annotations
from typing import Generic, TypeVar

T = TypeVar("T")
E = TypeVar("E")


class Result(Generic[T, E]):
    pass


class Ok(Result[T, E]):
    __match_args__ = ("value",)

    def __init__(self, value: T) -> None:
        self._value = value

    @property
    def value(self) -> T:
        return self._value


class Err(Result[T, E]):
    __match_args__ = ("error",)

    def __init__(self, error: E) -> None:
        self._error = error

    @property
    def error(self) -> E:
        return self._error


def process(result: Result[int, str]) -> int:
    match result:
        case Ok(value):
            reveal_type(value)  # expected: `int`, actual: `object`
            return value        # error: expected `int`, found `object`
        case Err(error):
            raise ValueError(error)  # error: cannot raise `object`
    raise AssertionError("unreachable")
```

**Command:** `ty check repro.py`

**Actual output:**
```
info[revealed-type]: Revealed type
 --> repro.py:37:25
  |
  |                         ^^^^^ `object`

error[invalid-return-type]: Return type does not match returned value
 --> repro.py:38:20
  |
  |                    ^^^^^ expected `int`, found `object`

error[invalid-raise]: Cannot raise object of type `object`
 --> repro.py:41:19
```

**Expected:** `value` inferred as `int` (the `T` of `Ok[int, str]`), `error` inferred as `str`.

The `__match_args__` entry `"value"` maps to the `value` property typed as `T`. When `result: Result[int, str]` is narrowed to `Ok[int, str]` by the class pattern, the positional capture should resolve `T = int`.

Related: #887

## Version

`ty 0.0.55 (42869400d 2026-06-26)`

**Comment by charliermarsh** [comment]
We added some support for this in https://github.com/astral-sh/ruff/pull/26411, e.g., when the specialization is on the subject:

```python
from typing import Generic, TypeVar

T = TypeVar("T")

class Base(Generic[T]):
    item: T

class Child(Base[T]):
    pass

def handle(value: Base[int]) -> None:
    match value:
        case Child(item=item):
            reveal_type(item)  # int
```

But not when it's on the subclass:

```python
class Base(Generic[T]):
    pass

class Child(Base[T]):
    item: T
    items: list[T]

def handle(value: Base[int]) -> None:
    match value:
        case Child(item=item, items=items):
            reveal_type(item)   # Unknown; should be int
            reveal_type(items)  # list[Unknown]; should be list[int]
```

**Comment by charliermarsh** [comment]
(Properties might also be harder and unsupported right now; I'd need to check, but either way there's a TODO here.)

**Comment by Jayllyz** [comment]
Thanks a lot ♥️
